### ISSUE TITLE
How would I add validation statistic summaries to a graph (and view in Tensorboard)?

### ISSUE THREAD

**Issue by skoppula** [ppwwyyxx/tensorpack#396]
Hi there!

Looking through one of the MNIST examples (e.g. https://github.com/ppwwyyxx/tensorpack/blob/master/examples/mnist-convnet.py), I see that add_moving_summary is included in the graph to track train_error and the regularization/cross entropy losses.

As such, when I open Tensorboard, I can see these train dataset statistics. This is great!

However, the validation summary statistics don't show up in Tensorboard. Validation statistics are being computed in the Callback:

```
InferenceRunner(    # run inference(for validation) after every epoch
               dataset_test,   # the DataFlow instance used for validation
                [ScalarStats('cross_entropy_loss'), ScalarStats('accuracy'),
                 ClassificationError('incorrect')]),
```

presumably because they aren't added as a summary variable in the graph.

Any idea how to add this in, so I can view validation statistics (as a function of epochs or interval'd steps)?

**Comment by skoppula** [comment]
The sort of lame way of doing validation tracking would be to scrape the log files which output the validation statistics at the end of every epoch -- is there a Tensorpack script to do this already?

**Comment by ppwwyyxx** [comment]
Validation statistics do go to tensorboard.

**Comment by ppwwyyxx** [comment]
Feel free to reopen if you have more questions.

**Comment by zxpeter** [comment]
Hi yx,

My question is similar and I'm using examples/FasterRCNN.
I didn't find any add_moving_summary() code in evaluation part, so I assume that's the reason why I can't see the Validation statistics on tensorboard.
![image](https://user-images.githubusercontent.com/12379916/56029348-c6758900-5d4c-11e9-99bb-b6ffac9333ac.png)
I also check up this issue #127, but not sure if it's right to modify like below.
`callbacks = [

            PeriodicCallback(
                ModelSaver(max_to_keep=10, keep_checkpoint_every_n_hours=1),
                every_k_epochs=20),
            # linear warmup
            ScheduledHyperParamSetter(
                'learning_rate', warmup_schedule, interp='linear', step_based=True),
            ScheduledHyperParamSetter('learning_rate', lr_schedule),
            PeakMemoryTracker(),
            EstimatedTimeLeft(median=True),
            SessionRunTimeout(60000).set_chief_only(True),   # 1 minute timeout
        ] + [
            EvalCallback(dataset, *MODEL.get_inference_tensor_names(), args.logdir)
            for dataset in cfg.DATA.VAL
        ] + [
            InferenceRunner(    # run inference(for validation) after every epoch
                val_dataflow,   # the DataFlow instance used for validation
                ScalarStats(    # produce `val_accuracy` and `val_cross_entropy_loss`
                    ['accuracy'], prefix='val'))
        ]
`

**Comment by ppwwyyxx** [comment]
>  InferenceRunner(    # run inference(for validation) after every epoch
            val_dataflow,   # the DataFlow instance used for validation
            ScalarStats(    # produce `val_accuracy` and `val_cross_entropy_loss`
                ['accuracy'], prefix='val'))
    ]

This is correct, except that there is no tensor named "accuracy" in the graph

Also note that the other callback `EvalCallback` also runs evaluation. So you'll run evaluation twice, with different metrics. You can also modify `EvalCallback` to add your own metrics.
